### PR TITLE
add more SubGHz support. add Silicon Labs NodeTest hardware support.

### PR DESCRIPTION
**Pull request by AdamLaurie**
This has been thoroughly tested with RZUSBSTICK and SiLabs NodeTest in 2.4GHz and 863MHz. I do not have access to any of the other sniffer hardware so I suggest we merge this and then sort those out if and when there are issues found (none were found with RZUSBSTICK so I'm not expecting any anywhere else).

### CODE REVIEW

File: killerbee/dev_apimote.py [hunk 1]
```diff
@@ -201,27 +209,29 @@ def pnext(self, timeout=100):
...
+    def ping(self, da, panid, sa, channel=None, page=0):
         '''
         Not yet implemented.
         @return: None
         @rtype: None
         '''
         raise Exception('Not yet implemented')
 
-    def jammer_on(self, channel=None):
+    def jammer_on(self, channel=Nonei, page=0):
```
[the remark this answers is not in the document]
Comment (AdamLaurie) [reply killerbee/dev_apimote.py]: Good spot, thanks! Fixed.

File: killerbee/dev_apimote.py [hunk 2]
```diff
@@ -122,9 +125,10 @@ def sniffer_off(self):
         self.__stream_open = False
 
     # KillerBee expects the driver to implement this function
-    def set_channel(self, channel):
+    def set_channel(self, channel, page=0):
         '''
-        Sets the radio interface to the specifid channel (limited to 2.4 GHz channels 11-26)
+        Sets the radio interface to the spec/
+ifid channel (limited to 2.4 GHz channels 11-26)
```
Comment (rmspeers) [inline killerbee/dev_apimote.py]: indentation issue?

Comment (AdamLaurie) [reply killerbee/dev_apimote.py]: thx! fixed!

File: killerbee/kbutils.py [hunk 3]
```diff
@@ -126,19 +126,22 @@ def is_valid_channel(self, channel, page= 0):
...
         @rtype: Boolean
         '''
-        # if sub-ghz, check that page and channel match
+        # if sub-ghz, check that page and channel and capability match
         if page:
-            if (page == 28 or page == 30 or page == 31) and channel > 26:
+            if page == 28 and (channel > 26 or not self.check(self.FREQ_863)):
                 return False
-            elif page == 29 and channel  > 8:
+            elif page == 29 and (channel > 8 or not self.check(self.FREQ_868)):
```
Comment (BroderickCarlin) [inline killerbee/kbutils.py]: You are switching between `if` and `elif` in this chain. It doesn't effect code flow at all, but might worth being consistent.

Comment (AdamLaurie) [reply killerbee/kbutils.py]: Yeah, that's what happens when you squeeze your own code in amongst someone else's! :) I prefer the raw 'if' myself... Thanks again for the spot.

File: killerbee/__init__.py [hunk 4]
```diff
@@ -140,6 +140,12 @@ def __init__(self, device=None, datasource=None, gps=None):
                 self.dev = device
                 if (self.dev == gps_devstring):
                     pass
+                elif (DEV_ENABLE_SL_NODETEST and kbutils.issl_nodetest(self.dev)):
+                    from dev_sl_nodetest import SL_NODETEST
+                    self.driver = SL_NODETEST(self.dev)
+                elif (DEV_ENABLE_SL_BEEHIVE and kbutils.issl_beehive(self.dev)):
+                    from dev_sl_beehive import SL_BEEHIVE
+                    self.driver = SL_BEEHIVE(self.dev)
```
[the remark this answers is not in the document]
Comment (AdamLaurie) [reply killerbee/__init__.py]: BeeHive is the version that can receive/inject Bad FCS - it's mentioned in the readme but not explicitly - just says "contact the author". I can't publish the source for either NodeTest of BeeHive as they are both under NDA.

File: killerbee/dev_sl_beehive.py [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+        self.capabilities.setcapab(KBCapabilities.SETCHAN, True)
+        return
+
+    def get_dev_info(self):
+        '''
+        Returns device information in a list identifying the device.
+        @rtype: List
+        @return: List of 3 strings identifying device.
+        '''
+        return [self.dev, "BeeHive SG", ""]
```
[the remark this answers is not in the document]
Comment (AdamLaurie) [reply killerbee/dev_sl_beehive.py]: I tried this with RZUSBSTICK but for some reason it didn't work. I agree, would be useful for next release!

File: killerbee/dev_sl_beehive.py [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+        self._page = page
+        if page:
+            channel = ((page << 5) & 0xff) + (channel & 0x1f)
+            if page == 31:
+                self.__send_cmd("915MHz", extra_delay= 0.3)
+            else:
+                self.__send_cmd("863MHz", extra_delay= 0.3)
+        else:
+            self.__send_cmd("2p4GHz")
+        self.__send_cmd("802154", "rx 100 192 864")
```
[the remark this answers is not in the document]
Comment (AdamLaurie) [reply killerbee/dev_sl_beehive.py]: Sure. I'll add some comments.

File: killerbee/dev_sl_nodetest.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+            self.set_channel(channel, page)
+
+        # Append two bytes to be replaced with FCS by firmware.
+        packet = ''.join([packet, "\x00\x00"])
+
+        raise Exception('Not yet implemented')
+
+        for pnum in range(0, count):
+            # Format for packet is opcode CMD_INJECT_FRAME, one-byte length,  packet data
+            #TODO RZ_USB_COMMAND_EP, struct.pack("B", RZ_CMD_INJECT_FRAME) + struct.pack("B", len(packet)) + packet)
```
[the remark this answers is not in the document]
Comment (AdamLaurie) [reply killerbee/dev_sl_nodetest.py]: Yeah, this was a massive edit as I had to change all the dev_ files so I probably missed a few things I should have tidied up... :)

File: killerbee/dev_sl_nodetest.py [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+
+        if not self.__stream_open:
+            # make sure we get hex dump of RX packet
+            self.__send_cmd("showPayload", "1")
+            # start sniffer
+            self.__send_cmd("rx", confirm= False)
+            for x in range(10):
+                d = self.handle.readline()
+                #print 'got', d
+                if '{payload}' in d:
```
[the remark this answers is not in the document]
Comment (AdamLaurie) [reply killerbee/dev_sl_nodetest.py]: No, this is a header that is output when RX is enabled. No packet needs to be received, so this is correct behaviour.

File: killerbee/dev_sl_beehive.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (AdamLaurie) [reply killerbee/dev_sl_beehive.py]: Doh. Turns out wasn't needed. Left over from earlier debugging.

File: killerbee/dev_sl_beehive.py [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+            return None   # Sense timeout case and return
+
+        #print packet
+        rssi, frame, validcrc = self.__dissect_pkt(packet)
+        if not frame:
+            print "Error parsing stream received from device:", packet
+
+        # Parse received data as <rssi>!<time>!<packtlen>!<frame>
+        try:
+            rssi = int(rssi)
```
[the remark this answers is not in the document]
Comment (AdamLaurie) [reply killerbee/dev_sl_beehive.py]: yep, done.

**Pull request merged**